Naïve Bayes classifier for the dataset

Import Statements

In [50]:
import pandas as pd
import numpy as np
from sklearn.naive_bayes import BernoulliNB, GaussianNB, MultinomialNB, ComplementNB
from sklearn.metrics import accuracy_score, precision_score
from sklearn.model_selection import train_test_split

importing the dataset

In [51]:
df = pd.read_csv("./CleanedData.csv", index_col=0)
columns = df.columns
print(df.head())

   GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  Research  \
0      337.0        118.0                  4  4.5   4.5  9.65         1   
1      324.0        107.0                  4  4.0   4.5  8.87         1   
2      316.0        104.0                  3  3.0   3.5  8.00         1   
3      322.0        110.0                  3  3.5   2.5  8.67         1   
4      314.0        103.0                  2  2.0   3.0  8.21         0   

   SES Percentage  Asian  african american  latinx  white  Chance of Admit  
0              12      1                 0       0      0             0.92  
1              11      0                 0       1      0             0.76  
2              78      0                 0       1      0             0.72  
3              77      0                 0       0      1             0.80  
4               1      0                 1       0      0             0.65  


define x

In [52]:
x = df[columns[0:-1]]

Again, like Decision tree, I have to binarize the y column. To do so, i must find the threshold, which for now will be
the median value

In [53]:
median_value = df["Chance of Admit"].median()
print(median_value)

0.73


0.73 is the median, and thus will use as the threshold for binarization.

In [54]:
for i in range(len(df["Chance of Admit"])):
    if df.loc[i, "Chance of Admit"] >= 0.73:
        df.loc[i, "Chance of Admit"] = 1
    else:
        df.loc[i, "Chance of Admit"] = 0
print(df["Chance of Admit"])

0      1.0
1      1.0
2      0.0
3      1.0
4      0.0
      ... 
351    1.0
352    1.0
353    1.0
354    0.0
355    1.0
Name: Chance of Admit, Length: 356, dtype: float64


Defining y


In [55]:
y = df["Chance of Admit"]

splitting x and y into training and testing data

In [56]:
x_train, x_test, y_train, y_test = train_test_split(x, y, shuffle=True, random_state=1)
print(x_train.head())
print()
print(y_train.head())

     GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  Research  \
23       300.0         97.0                  2  3.0   3.0  8.10         1   
320      338.0        115.0                  5  4.5   5.0  9.23         1   
213      328.0        110.0                  4  4.0   2.5  9.02         1   
51       307.0        101.0                  3  4.0   3.0  8.20         0   
92       301.0        107.0                  3  3.5   3.5  8.34         1   

     SES Percentage  Asian  african american  latinx  white  
23               93      0                 0       1      0  
320              19      0                 0       0      1  
213              15      0                 0       0      1  
51               94      0                 0       0      1  
92               52      0                 0       1      0  

23     0.0
320    1.0
213    1.0
51     0.0
92     0.0
Name: Chance of Admit, dtype: float64


Creating Gaussian, Bernoulli, Multinominal, and Complement Naïve Bayes learners.

In [57]:
gaussian_model = GaussianNB()
bernoulli_model = BernoulliNB()
multinominal_model = MultinomialNB()
complement_model = ComplementNB()


Fitting each model with the training data

In [58]:
gaussian_model.fit(x_train, y_train)
bernoulli_model.fit(x_train, y_train)
multinominal_model.fit(x_train, y_train)
complement_model.fit(x_train, y_train)


ComplementNB()

Obtaining y_pred from each model


In [59]:
y_pred_g = gaussian_model.predict(x_test)
y_pred_b = bernoulli_model.predict(x_test)
y_pred_m = multinominal_model.predict(x_test)
y_pred_c = complement_model.predict(x_test)

Print accuracy and precision scores

In [62]:
print("Gaussian Model accuracy: " + str(accuracy_score(y_test, y_pred_g)))
print("Gaussian Model precision: " + str(precision_score(y_test, y_pred_g)))
print("Bernoulli Model accuracy: " + str(accuracy_score(y_test, y_pred_b)))
print("Bernoulli Model precision: " + str(precision_score(y_test, y_pred_b)))
print("Multinominal Model accuracy: " + str(accuracy_score(y_test, y_pred_m)))
print("Multinominal Model precision: " + str(precision_score(y_test, y_pred_m)))
print("Complement Model accuracy: " + str(accuracy_score(y_test, y_pred_c)))
print("Complement Model precision: " + str(precision_score(y_test, y_pred_c)))

Gaussian Model accuracy: 0.8426966292134831
Gaussian Model precision: 0.8627450980392157
Bernoulli Model accuracy: 0.7752808988764045
Bernoulli Model precision: 0.803921568627451
Multinominal Model accuracy: 0.5617977528089888
Multinominal Model precision: 0.62
Complement Model accuracy: 0.5842696629213483
Complement Model precision: 0.6346153846153846


It appears the Gaussian Naïve Bayes is the best predictor of the Bayes classifiers, with an accuracy of 84.27% and a
precision of 86.27%

Naïve Bayes is a relatively simple learner. I was surpised at the classification accuracy it was able to achieve. However,
I feel slightly powerless because there are not many hyperparameters to alter, except for only the types of NB classifiers.